cpp/Json.hpp

```diff
@@ -0,0 +1,405 @@
+#pragma once
+
+/// @brief JSON stream parser
+
+#include <queue>
+#include <cstring>
+
+#include "Types.hpp"
+#include "String.hpp"
+#include "Result.hpp"
+#include "Vector.hpp"
+#include "UniquePtr.hpp"
+#include "UnorderedMap.hpp"
+
+namespace hsd {
+    enum class JsonToken {
+        Null, True, False,
+        Number, String,
+        BArray, EArray, BObject, EObject,
+        Comma, Colon,
+
+        Eof = -1,
+        Error = -2,
+        Empty = -3 // For internal purposes
+    };
+
+    class JsonError : public runtime_error {
+    public:
+        usize position;
+
+        JsonError(const char* msg, usize pos)
+            : runtime_error(msg), position(pos) {}
+    };
+
+    template <typename CharT>
+    class JsonStream {
+        using vstr = basic_string_view<CharT>;
+        using str = basic_string<CharT>;
+
+        std::queue<JsonToken> tokens;
+        std::queue<str> qtok_string;
+        std::queue<i32> qtok_number;
+        JsonToken current_token = JsonToken::Empty;
+        const char* token_kw = nullptr;
+        usize token_position = 0;
+        str token_str;
+
+        usize pos = 0;
+
+    public:
+        std::queue<JsonToken>& get_tokens() {return tokens; }
+        str pop_string() {
+            str s = move(qtok_string.front());
+            qtok_string.pop();
+            return s;
+        }
+        i32 pop_number() {
+            i32 n = qtok_number.front();
+            qtok_number.pop();
+            return n;
+        }
+
+        // If an error occurs, the rest of buffer can be passed in
+        Result<void, JsonError> lex(vstr frag) {
+            static const char* const s_keywords[] = {"null", "true", "false"};
+            for (CharT ch : frag) {
+                ++pos;
+                _reparse:
+                if (current_token == JsonToken::Empty) {
+                    if (std::isspace(ch))
```

Comment:
Explain this?

Reply:
To skip whitespace, duh

Reply:
You couldn't do `if (ch == static_cast<CharT>(' '))`?

Reply:
Do you know how many whitespace characters are there?

Reply:
m a n y